File: data-structure-algorithm/160-Intersection_of_Two_Linked_Lists.js
/**
 * 
 * 
 *  160. Intersection of Two Linked Lists
 * 
    Write a program to find the node at which the intersection of two singly linked lists begins.

    求两个链表的交叉点

    A: a1 → a2 
    ↘ 
    c1 → c2 → c3 
    ↗ 
    B: b1 → b2 → b3 


    a1 a2 c1 c2 c3 b1 b2 b3 c1 c2 c3
    b1 b2 b3 c1 c2 c3 a1 a2 c1 c2 c3

 * 
 */

 /**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */

var getIntersectionNode = function(headA, headB) {
  var map = new Map()
  while (heapA) {
    // temp.push(tempA)
    map.set(headA, 1)
    headA = headA.next
  }

  while(headB) {
    if(map.has(headB)) {
      return headB
    }
    headB = headB.next
  }
  return null
};

var getIntersectionNode = function(heapA, heapB) {
  let p1 = heapA
  let p2 = heapB
  while (p1 != p2){
    p1 = p1 == undefined ? heapB : p1.next
    p2 = p2 == undefined ? heapA : p2.next
  }
  return p1
}

File: data-structure-algorithm/golang/160. Intersection of Two Linked Lists.go
package ds

/**



 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

// 方法一，用Set或者Map存储, Set没有原生GO，用Map操作
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	m := make(map[*ListNode]int)

	for headA != nil {
		m[headA] = 1
		headA = headA.Next
	}

	for headB != nil {
		_, ok := m[headB]
		if ok {
			return headB
		}
		headB = headB.Next
	}
	return nil
}

// 方法二 略 a+c+b = b+c+a
// A遍历完略找B，B遍历完了找A
